flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/SingleLogicalSlotTest.java

```diff
@@ -109,6 +109,23 @@ public void testPayloadAssignmentAfterRelease() {
 		assertThat(singleLogicalSlot.tryAssignPayload(dummyPayload), is(false));
 	}
 
+	@Test
+	public void testReleaseOnlyHappensAfterReturningSlotCompleted() {
+
+		final SingleLogicalSlot singleLogicalSlot = createSingleLogicalSlot(new DummySlotOwner() {
+			@Override
+			public CompletableFuture<Boolean> returnAllocatedSlot(LogicalSlot logicalSlot) {
+				// we return a future that will never complete instead
+				return new CompletableFuture<>();
+			}
+		});
+
+		CompletableFuture<?> releaseFuture = singleLogicalSlot.releaseSlot(new FlinkException("Test exception"));
+
+		// this must not be done, because the future that signals a successful return never completed.
+		assertThat(releaseFuture.isDone(), is(false));
```

Comment:
It would be good to also test that if we now complete the `returnAllocatedSlot` future that the `releaseFuture` will also be completed.